src/ir_passes/new_mono.rs

```diff
@@ -0,0 +1,868 @@
+        let bound = mono_end.as_concrete(&self.base).unwrap();
+
+        while i < bound {
+            self.binding.insert(*index, i);
+            for cmd in body.iter() {
+                let cmd = self.command(cmd);
+                self.base.cmds.extend(cmd);
+            }
+            i += 1;
+        }
+    }
+
+    fn if_stmt(&mut self, if_stmt: &ir::If) {
+        let ir::If { cond, then, alt } = if_stmt;
+
+        let cond = self.prop(*cond);
+        let cond = self.base.resolve_prop(self.base.get(cond).clone());
+
+        let then = then
+            .iter()
+            .map(|cmd| self.command(cmd))
+            .fold(&mut vec![], |acc, cvec| {
+                acc.extend(cvec);
+                acc
+            })
+            .to_vec();
+        let alt = alt
+            .iter()
+            .map(|cmd| self.command(cmd))
+            .fold(&mut vec![], |acc, cvec| {
+                acc.extend(cvec);
+                acc
+            })
+            .to_vec();
+
+        match cond {
+            ir::Prop::True => self.base.cmds.extend(then),
+            ir::Prop::False => self.base.cmds.extend(alt),
+            _ => panic!("couldnt resolve prop"),
+        }
+    }
+
+    fn fact(&mut self, fact: &ir::Fact) -> Option<ir::Command> {
+        let ir::Fact { prop, reason, .. } = fact;
+
+        let prop = self.prop(*prop);
+        let reason = self.info(reason);
+        if fact.is_assert() {
+            self.base.assert(prop, reason)
+        } else {
+            self.base.assume(prop, reason)
+        }
+    }
+
+    fn eventbind(&mut self, eb: &ir::EventBind) -> ir::EventBind {
+        let ir::EventBind {
+            arg,
+            info,
+            delay,
+            base,
+        } = eb;
+
+        let delay = self.timesub(delay);
+        let arg = self.time(*arg);
+        let info = self.info(info);
+        let base = self.foreign(base);
+
+        ir::EventBind {
+            arg,
+            info,
+            delay,
+            base,
+        }
+    }
+
+    /// Either queues a connect to be handled later, or handles it now.
+    /// A connect gets handled later if:
+    /// (1) The source port is invoke-owned and the destination port is also invoke-owned, or
+    /// (2) The source port is sig-owned and the destination port is invoke-owned
+    /// A connect gets handled now if:
+    /// (1) The source port is invoke-owned and the destination port is sig-owned
+    /// (2) The source port is sig-owned and the destination port is also sig-owned
+    /// The reasoning for this is that we see connects where the destination port is invoke-owned before
+    /// we see the actual invoke, so we have to put them in a queue to be handled when we see the invoke
+    /// If the destination port is not invoke owned, then we can handle it now because we've already created
+    /// the monomorphized ports
+    fn handle_connect(&mut self, con: &ir::Connect) -> Vec<ir::Command> {
```

Comment:
Invokes are stored in a separate table so can you iterate over them before walking over the commands and get rid of this delayed handling?

Reply:
If you have something like `f := new Foo<G>(in)` the commands are laid out as:
```
inst0
p1 <- p0
inv0
```
so when I'm processing a component I just iterate through the commands, meaning I'll see the connects before the invokes. I would iterate over the invokes first to generate them, but in the case where there is a loop in the commands I don't think this would work, because if you have an invoke in a loop the only thing that is in `underlying` is that singular invoke. So my thought was that as we unroll loops, we generate the invokes and then connect their ports

Reply:
I see! We can try to make it an IR invariant that things must get defined before they are used. For example, the `invoke` statement defines the input and output ports:
```
inv, p1 = invoke
p1 <- p0
```
This means, it should come before the port connection in the control program. This is usually called the domination condition where the definition of something must dominate its use.

Reply:
This invariant could also be enforced using the validation pass.